eclair-core/src/main/scala/fr/acinq/eclair/payment/Bolt12Invoice.scala

```diff
@@ -39,57 +39,34 @@ case class Bolt12Invoice(records: TlvStream[InvoiceTlv]) extends Invoice {
-  val payerKey: Option[ByteVector32] = records.get[PayerKey].map(_.publicKey)
-  val payerNote: Option[String] = records.get[PayerNote].map(_.note)
-  val payerInfo: Option[ByteVector] = records.get[PayerInfo].map(_.info)
-  val fallbacks: Option[Seq[FallbackAddress]] = records.get[Fallbacks].map(_.addresses)
-  val refundSignature: Option[ByteVector64] = records.get[RefundSignature].map(_.signature)
-  val replaceInvoice: Option[ByteVector32] = records.get[ReplaceInvoice].map(_.paymentHash)
+  override val nodeId: Crypto.PublicKey = records.get[InvoiceNodeId].get.nodeId
+  override val paymentHash: ByteVector32 = records.get[InvoicePaymentHash].get.hash
+  override val description: Either[String, ByteVector32] = Left(invoiceRequest.offer.description)
+  override val createdAt: TimestampSecond = records.get[InvoiceCreatedAt].get.timestamp
+  override val relativeExpiry: FiniteDuration = FiniteDuration(records.get[InvoiceRelativeExpiry].map(_.seconds).getOrElse(DEFAULT_EXPIRY_SECONDS), TimeUnit.SECONDS)
+  override val minFinalCltvExpiryDelta: CltvExpiryDelta = CltvExpiryDelta(0)
+  override val features: Features[InvoiceFeature] = records.get[InvoiceFeatures].map(_.features.invoiceFeatures()).getOrElse(Features.empty)
+  val blindedPaths: Seq[RouteBlinding.BlindedRoute] = records.get[InvoicePaths].get.paths
+  val blindedPathsInfo: Seq[PaymentInfo] = records.get[InvoiceBlindedPay].get.paymentInfo
+  val fallbacks: Option[Seq[FallbackAddress]] = records.get[InvoiceFallbacks].map(_.addresses)
   val signature: ByteVector64 = records.get[Signature].get.signature
 
   // It is assumed that the request is valid for this offer.
-  def isValidFor(offer: Offer, request: InvoiceRequest): Boolean = {
-    nodeId == offer.nodeId &&
-      checkSignature() &&
-      offerId.contains(request.offerId) &&
-      request.chain == chain &&
+  def isValidFor(request: InvoiceRequest): Boolean = {
+    invoiceRequest.unsigned == request.unsigned &&
+      nodeId == invoiceRequest.offer.nodeId &&
       !isExpired() &&
-      request.amount.contains(amount) &&
-      quantity == request.quantity_opt &&
-      payerKey.contains(request.payerKey) &&
-      payerInfo == request.payerInfo &&
-      // Bolt 12: MUST reject the invoice if payer_note is set, and was unset or not equal to the field in the invoice_request.
-      payerNote.forall(request.payerNote.contains(_)) &&
-      description.swap.exists(_.startsWith(offer.description)) &&
-      issuer == offer.issuer &&
-      request.features.areSupported(features)
-  }
-
-  def checkRefundSignature(): Boolean = {
-    (refundSignature, refundFor, payerKey) match {
-      case (Some(sig), Some(hash), Some(key)) => verifySchnorr(signatureTag("payer_signature"), hash, sig, key)
-      case _ => false
-    }
+      request.amount.forall(_ == amount) &&
+      request.features.areSupported(features) &&
+      checkSignature()
   }
 
   def checkSignature(): Boolean = {
```

Comment:
Does it really make sense to have a `checkSignature` method (and same comment for `InvoiceRequest`)? Shouldn't we just verify it in the `validate` functions and then stop caring about it? Otherwise we just risk forgetting to verify it.

Reply:
It is checked in the `isValidFor`.

Reply:
Linked to other comments, I believe all of those should be done in `validate`, it's confusing and error-prone to scatter validation across different methods that must be called separately, isn't it?